# CircRM
## Brief introduction

CircRM is a computational tool for profiling circular RNAs (circRNAs) using Nanopore direct RNA sequencing (DRS) data. It can detect circRNAs and identify RNA modifications (m⁶A, m¹A, and m⁵C) at single-molecule, single-base resolution. Designed for flexibility, circRM works across multiple cell lines and species, enabling comprehensive analysis of circRNA epitranscriptomes.
## Installation
```bash
git clone https://github.com/jiayiAnnie17/CircRM
```
## 1 Profiling of circRNAs
### 1.1 Prepare virtual library

```bash
Rscript ./CircRM/Code/Build_VL.R <TxDb_annotation> <output_csv>
```
#### Dependencies

Make sure the following R packages are installed before running the script:  
`GenomicRanges`, `GenomicFeatures`, `rtracklayer`, `data.table`, `tidyr`

#### Parameter explaination
- `<TxDb_annotation>`: a transcript annotation object from a TxDb package.
  Examples:
   - `TxDb.Mmusculus.UCSC.mm10.knownGene` for mouse
   - `TxDb.Hsapiens.UCSC.hg38.knownGene` for human

- `<output_csv>`: the CSV file generated by this script

The script accepts any species and genome assembly for which a TxDb annotation package is available; just provide the appropriate TxDb object. 

#### Output File Format

The output file contains six columns as follows:

| Column | Description |
|--------|-------------|
| `Transcript_ID:Location` | Transcript ID and genomic coordinates (e.g., `ENST00000456328.2chr1:11869-12227`) |
| `Chromosome` | Chromosome name (e.g., `chr1`) |
| `Start` | Start position of the transcript or feature (e.g., `11869`) |
| `End` | End position of the transcript or feature (e.g., `12227`) |
| `Strand` | Strand information (`+` or `-`) |
| `Transcript_ID` | Transcript ID (e.g., `ENST00000456328.2`) |


### 1.2 Split
This step groups the input CSV by transcript, enabling efficient processing of large files.

```bash
python ./CircRM/Code/Split.py -i <input_csv> -o <output_directory> [-b <batch_size>]
```
#### Parameter explaination
- `-i`: the CSV file generated by the previous step

- `-o`: the folder where the output files will be saved

- `-b` : optional, the number of transcripts per batch (default: 50)
### 1.3 Find circRNAs
如果没有fastq可以使用下面2.2 guppy生成pass的fastq再来这里
```bash
python find_circRNA.py -i <input_fastq> -d <csv_directory> -r <reference_fasta> -o <output_file> [-t <threads>] [-m <merged_fasta>]
```
This script requires the following Python packages: `os`, `pandas`, `argparse`, `sys`, `tqdm`, `biopython` (for `Bio`), and `mappy`.

#### Parameters explaination

- `-i` : Path to the input FASTQ file containing sequencing reads
- `-d` : Path to the directory containing CSV files required for processing
- `-r` : Reference genome FASTA file (e.g., hg38.fa)
- `-o` : Output file to save circRNA detection results
- `-t` : Number of threads to use (default: 1)
- `-m` : Output FASTA file name for merged sequences (default: `merged_sequences.fasta`)
#### Output File Format

The output file contains three columns:

| Column | Description |
|--------|-------------|
| `Read_ID` | Unique identifier of the read (e.g., `8419b769-3b61-422a-9188-543e793680a2`) |
| `Transcript_Location` | Transcript ID and genomic coordinates (e.g., `ENST00000459737.5chr14:55650328-55659703`) |
| `Strand` | Original strand of the reference sequence (`+` or `-`) |

All circRNAs are identified in this step; in general, those with coverage **above 30** are regarded as high-confidence.
## 2 Find modifications
### 2.1 Retrieve circRNA Raw Signals from FAST5

In this step, we use `fast5_subset.py` to extract FAST5 files containing circRNA reads identified in previous steps.

```bash
python fast5_subset.py \
  -i <input_fast5_dir> \
  -s <output_dir> \
  -l <read_id_list> \
  -r \
  -t <threads>
```
#### Parameters explaination in this workflow

- `-i` : Input directory containing the **original raw FAST5 files**.  
- `-s` : Output directory where **circRNA FAST5 files** will be saved.  
- `-l` : Text file containing **read IDs corresponding to circRNAs**.  
- `-r` : Flag to retain the original FAST5 folder structure.   
- `-t` : Number of threads to use (default: 1).  

### 2.2 guppy
### multi
### tombo
### extract feature
### train
### predict 
### Methylation level
默认是30和5，可以自己改
